# Task 4.6

# 1. Import libraries

In [33]:
import pandas as pd
import numpy as np
import os

# 2. Set Path

In [34]:
# set path for data files
path = r'C:\Users\Stony\OneDrive\CareerFoundry\Data Immersion Course\Instacart Basket Analysis'

# 3. Import data sets

In [35]:
# Import orders_products_combined.pkl into a dataframe
df_ords_prods_combined = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

In [36]:
# Import products_cleaned into a dataframe
df_prods_clean = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_cleaned.csv'), index_col = False)

# 4. Check the dimensions of the imported dataframes

In [37]:
# Check dimensions of df_ords_prods_combined
df_ords_prods_combined.shape

(32434489, 11)

In [38]:
df_ords_prods_combined.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_last_order  first_order  product_id  add_to_cart_order  \
0                    NaN         True         196                  1   
1                    NaN         True       14084                  2   
2                    NaN         True       12427                  3   
3                    NaN         True       26088                  4   
4                    NaN         True       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

In [39]:
# Drop unneeded _merge column from ords_prods_combined dataframe
df_ords_prods_combined = df_ords_prods_combined.drop(['_merge'], axis=1)

In [40]:
# Verify column was dropped and confirm shape is still consistent
df_ords_prods_combined.shape

(32434489, 10)

In [41]:
df_ords_prods_combined.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_last_order  first_order  product_id  add_to_cart_order  \
0                    NaN         True         196                  1   
1                    NaN         True       14084                  2   
2                    NaN         True       12427                  3   
3                    NaN         True       26088                  4   
4                    NaN         True       26405                  5   

   reordered  
0          0  
1          0  
2          0  
3          0  
4          0

In [42]:
# Check dimensions of df_prods dataframe
df_prods_clean.shape

(49672, 6)

In [43]:
df_prods_clean.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [44]:
# Drop unneeded column from products_cleaned dataframe
df_prods_clean = df_prods_clean.drop(['Unnamed: 0'], axis=1)

In [45]:
df_prods_clean.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [46]:
df_prods_clean.shape

(49672, 5)

# 5. Merge prods_clean and ords_prods_combined dataframes on product_id column

In [47]:
df_ords_prods_merged = df_prods_clean.merge(df_ords_prods_combined, on = 'product_id', indicator = True)

In [48]:
# Validate merge was successful
df_ords_prods_merged.shape

(32404859, 15)

In [50]:
df_ords_prods_merged.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   3139998      138            28                  6                 11   
1   1977647      138            30                  6                 17   
2    389851      709             2                  0                 21   
3    652770      764             1                  3                 13   
4   1813452      764             3                  4                 17   

   days_since_last_order  first_order  add_to_cart_order  reordered _merge  
0                    3.0        False                  5          0   both  
1                   20.0        False                  1          1   both  
2                    6.0        False                 20          0   both  
3                    NaN         True                 10          0   both  
4                    9.0        False                 11          1   both

# 6. Confirm results of merge

In [51]:
df_ords_prods_merged['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

Using the merge flag, we can see that there are 32,404,859 rows that have data in both of the data sets, which makes sense since we used the default setting of an inner join

# 7. Export merged dataframe

In [ ]:
# Export df_ords_prods_merged to pickle 
df_ords